docs/CInterop.md

```diff
@@ -0,0 +1,165 @@
+# C Interop
+
+This is an extension of what is was covered in the [Language Guide](./LanguageGuide.md#c-interop).
+
+## Managed types
+
+In Carp types like `String` and `Array` are _managed_ types in that they are
+allocated on the Heap and the compiler will automatically free the allocated
+memory when they go out of scope. We'll see how we can go from these managed
+type to C and back.
+
+### String
+
+To use a managed `String` with a C function requiring a `char*` you can use the
+`String.cstr` function that will turn your `(Ref String)` into `(Ptr CChar)`:
+
+
+```clojure
+(register puts (Fn [(Ptr CChar)] ()))
+
+(let [a-str @"A string."]
+  (puts (String.cstr &a-str)))
+(puts (String.cstr "Hello"))
+```
+
+It might be a good idea not to expose the C type to the end user directly:
+
+```clojure
+(defmodule MyMod
+  (hidden puts-c)
+  (private puts-c)
+  (register puts-c (Fn [(Ptr CChar)] ()) "puts")
+  (defn puts [str-ref] (puts-c (String.cstr str-ref))))
+
+(let [a-str @"A string."]
+  (MyMod.puts &a-str))
+(MyMod.puts "Hello")
+```
+
+---
+
+If you are given a `char*` and want to turn it into a managed `String` you can
+use `String.from-cstr`. It will allocate and copy the content of the C string.
+
+```c
+// static-str.h
+char* returns_a_static_str() {
+  return "Hello";
+}
+```
+
+```clojure
+(relative-include "static-str.h")
+
+(register returns-a-static-str (Fn [] (Ptr CChar)) "returns_a_static_str")
+
+(let [a-str (String.from-cstr (returns-a-static-str))]
+  (println* (String.concat &[a-str @" " @"Carp"])))
+```
+
+---
+
+The function you're consuming might be allocating the string on the Heap for
+you. In that case you can declare the function as returning a managed `String`.
+However this might be unsafe, you need to ensure that the string is actually
+Heap-allocated and that the allocator is the same as the one that Carp is
+using.
+
+```c
+char* returns_a_heap_string() {
+  char *hello = "Hello from the heap";
+  char *str = malloc((strlen(hello)+1));
+  strcpy(str, hello);
+  return str;
+}
+```
+
+```clojure
+(relative-include "heap-string.h")
+
+(register returns-a-heap-str (Fn [] String) "returns_a_heap_string")
+
+(let [a-str (returns-a-heap-str)]
+  (println* a-str))
+```
+
+If you are the one writing the C code, you can use the `CARP_MALLOC` macro to
+ensure you are using the same allocator as the Carp compiler:
+
+```c
+char* returns_a_heap_string() {
+  char *hello = "Hello from the heap";
+  char *str = CARP_MALLOC((strlen(hello)+1));
```

Comment:
Wonder if we should use `String` here as Carp does internally https://github.com/carp-lang/Carp/blob/master/core/carp_string.h

Reply:
Although for the purposes of this document maybe the current state is better

Reply:
These functions are meant as examples of functions from an external library, would be unlikely for them to have `String` as the return type.
I might mention it in the next section about wrapping functions in custom C code.